core/src/com/etheller/warsmash/parsers/jass/Jass2.java

```diff
@@ -6280,644 +4055,364 @@ public static void registerTypingNatives(final JassProgramVisitor jassProgramVis
 			}
+			if (end < start) {
+				end = start;
+			}
+			return new StringJassValue(s.substring(start, end));
 		});
-		jassProgramVisitor.getJassNativeManager().createNative("GetLocalizedString", new JassFunction() {
-			@Override
-			public JassValue call(final List<JassValue> arguments, final GlobalScope globalScope,
-					final TriggerExecutionScope triggerScope) {
-				final String key = arguments.get(0).visit(StringJassValueVisitor.getInstance());
-				// TODO this might be wrong, or a subset of the needed return values
-				final String decoratedString = gameUI.getTemplates().getDecoratedString(key);
-				if (key.equals(decoratedString)) {
-					System.err.println("GetLocalizedString: NOT FOUND: " + key);
-				}
-				return new StringJassValue(decoratedString);
+		jassProgramVisitor.getJassNativeManager().createNative("StringLength", (arguments, globalScope, triggerScope) -> {
+			final String s = arguments.get(0).visit(StringJassValueVisitor.getInstance());
+			return new IntegerJassValue(s.length());
+		});
+		jassProgramVisitor.getJassNativeManager().createNative("StringCase", (arguments, globalScope, triggerScope) -> {
+			final String s = arguments.get(0).visit(StringJassValueVisitor.getInstance());
+			final boolean upper = arguments.get(1).visit(BooleanJassValueVisitor.getInstance());
+			return new StringJassValue(upper ? s.toUpperCase(Locale.US) : s.toLowerCase(Locale.US));
+		});
+		jassProgramVisitor.getJassNativeManager().createNative("StringHash", (arguments, globalScope, triggerScope) -> {
+			final String s = arguments.get(0).visit(StringJassValueVisitor.getInstance());
+			return new IntegerJassValue(s.hashCode());
+		});
+		jassProgramVisitor.getJassNativeManager().createNative("GetLocalizedString", (arguments, globalScope, triggerScope) -> {
+			final String key = arguments.get(0).visit(StringJassValueVisitor.getInstance());
+			// TODO this might be wrong, or a subset of the needed return values
+			final String decoratedString = gameUI.getTemplates().getDecoratedString(key);
+			if (key.equals(decoratedString)) {
+				System.err.println("GetLocalizedString: NOT FOUND: " + key);
 			}
+			return new StringJassValue(decoratedString);
 		});
-		jassProgramVisitor.getJassNativeManager().createNative("GetLocalizedHotkey", new JassFunction() {
-			@Override
-			public JassValue call(final List<JassValue> arguments, final GlobalScope globalScope,
-					final TriggerExecutionScope triggerScope) {
-				final String key = arguments.get(0).visit(StringJassValueVisitor.getInstance());
-				// TODO this might be wrong, or a subset of the needed return values
-				final String decoratedString = gameUI.getTemplates().getDecoratedString(key);
-				if (key.equals(decoratedString)) {
-					System.err.println("GetLocalizedHotkey: NOT FOUND: " + key);
-				}
-				return new IntegerJassValue(decoratedString.charAt(0));
+		jassProgramVisitor.getJassNativeManager().createNative("GetLocalizedHotKey", (arguments, globalScope, triggerScope) -> {
```

Comment:
Prior to your change, `GetLocalizedHotkey` had a lowercase 'k' in `key`, but after the change there is a typo making it `GetLocalizedHotKey`. Unfortunately, this string is case sensitive in the current Warsmash implementation and matched against the contents of possible custom maps that already exist with the lowercase. So, unless we were going to reinvent the JassNativeManager's "name to function" map to be case insensitive, this name should not be changed and should remain as the original `GetLocalizedHotkey` to retain functionality.